File: 0322-coin-change/0322-coin-change.go
/*
import (
    "math"
)

func coinChange(coins []int, amount int) int {
    table := make([]int, amount+1)
    for i := 1; i < len(table); i++ {
        minCoins := math.MaxInt32
        for _, c := range coins {
            if i-c >= 0 && table[i-c] != math.MaxInt32 {
                minCoins = min(minCoins, 1+table[i-c])
            }
        }
        table[i] = minCoins
    }
    if table[amount] == math.MaxInt32 {
        return -1
    }
    return table[amount]
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
*/

func coinChange(coins []int, amount int) int {
    cache := make(map[int]int)
    
    var coinChangeRecursive func([]int, int) int
    coinChangeRecursive = func(coins []int, amount int) int {
        if v, ok := cache[amount]; ok {
            return v
        }
        if amount == 0 {
            return 0
        }
        minCoins := math.MaxInt32
        for _, c := range coins {
            if amount-c >= 0 {
                minCoins = min(minCoins, 1+coinChangeRecursive(coins, amount-c))
            }
        }
        cache[amount] = minCoins
        return minCoins
    }
    
    res := coinChangeRecursive(coins, amount)
    if res == math.MaxInt32 {
        return -1
    }
    return res
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: 0322-coin-change/0322-coin-change.py
MAX_INT = 2**32

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        self.cache = {0: 0}
        min_coins = self.coin_change_recursive(coins, amount)
        if min_coins == MAX_INT:
            return -1
        return min_coins

    def coin_change_recursive(self, coins: List[int], amount: int) -> int:
        if self.cache.get(amount) != None:
            return self.cache[amount]
        if amount in coins:
            self.cache[amount] = 1
            return 1
        
        min_coins = MAX_INT
        for c in coins:
            if amount-c >= 0:
                min_coins = min(min_coins, 1+self.coin_change_recursive(coins, amount-c))
                
        self.cache[amount] = min_coins
        return min_coins            
"""
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        MAX_INT = float('inf')
        table = [0] + [MAX_INT] * (amount)
        
        for i in range(1, len(table)):
            min_coins = MAX_INT
            for c in coins:
                if i-c == 0:
                    min_coins = min(min_coins, 1)
                elif i-c > 0 and table[i-c] != MAX_INT:
                    min_coins = min(min_coins, 1+table[i-c])
            table[i] = min_coins
        
        if table[amount] == MAX_INT:
            return -1
        return table[amount]
"""